src/Transform/OptimizeMemoryPools.cpp

```diff
@@ -0,0 +1,905 @@
+  // of the firstGetRefList.
+  for (auto firstGetRef : firstGetRefList) {
+    // Fetch the full live range for the first set of getref operations.
+    std::vector<Operation *> liveRangeOpList = getLiveRange(firstGetRef);
+
+    // Get first and last ops for the live range of the secondGetRef.
+    Operation *firstOp = getLiveRangeFirstOp(secondGetRef);
+    Operation *lastOp = getLiveRangeLastOp(secondGetRef);
+
+    // Check if either the first or last ops in the second live range are part
+    // of the first live range.
+    bool firstOpInLiveRange = operationInLiveRange(firstOp, liveRangeOpList);
+    bool lastOpInLiveRange = operationInLiveRange(lastOp, liveRangeOpList);
+    if (firstOpInLiveRange || lastOpInLiveRange)
+      return true;
+
+    // Since firstOp and lastOp are not part of the live range, check whether
+    // the live range is fully contained between firstOp and lastOp. If it is
+    // return true.
+    if (liveRangeIsContained(firstOp, lastOp, liveRangeOpList))
+      return true;
+
+    // Up to this point, the checks we have done allow for ranges to be
+    // considered disjoint even when their extremities are part of the same
+    // loop nest. This means we have to perform an additional check: if the
+    // extremities of the two live ranges share the same loop nest determiend
+    // by `krnl.iterate` ops. If they do then the live ranges intersect.
+    if (liveRangesInSameLoopNest(firstOp, lastOp, liveRangeOpList))
+      return true;
+  }
+
+  // If all getRef live ranges are independent then no intersection exists.
+  return false;
+}
+
+//===----------------------------------------------------------------------===//
+// Rewrite patterns.
+//===----------------------------------------------------------------------===//
+
+class KrnlOptimizeStaticMemoryPools : public OpRewritePattern<KrnlGetRefOp> {
+public:
+  using OpRewritePattern<KrnlGetRefOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(
+      KrnlGetRefOp firstGetRef, PatternRewriter &rewriter) const override {
+    auto loc = firstGetRef.getLoc();
+    auto memRefType = convertToMemRefType(firstGetRef.getResult().getType());
+    auto memRefShape = memRefType.getShape();
+
+    // Only handle krnl.getref ops that return a constant shaped MemRef.
+    if (!hasAllConstantDimensions(memRefType))
+      return failure();
+
+    // Retrieve the AllocOp that this GetRef uses.
+    auto staticMemPool = getAllocOfGetRef(&firstGetRef);
+
+    // Ensure that the alloc obtained above is static memory pool.
+    auto memPoolType = convertToMemRefType(staticMemPool.getResult().getType());
+    auto memPoolShape = memPoolType.getShape();
+
+    // Static memory pool type must be byte.
+    if (getMemRefEltSizeInBytes(memPoolType) != 1)
+      return failure();
+
+    // Rank of the static memory pool must be 1.
+    if (memPoolShape.size() != 1)
+      return failure();
+
+    // Determine if the static memory pool is bundled i.e. participates in more
```

Comment:
if a, b, and c have mutually disjoint liveranges, could they share the same memory? In principle they can.. say you merge a&b. Then the liveranges becomes the union of their respective ranges.

This is where doing graph coloring is useful. You determine the interferences of the liveranges, and then you can merge things without "thinking" about the original loads/store, because all info is incapsulated into the graph.

Reply:
Yes they can and will share the same slot provided they satisfy all the conditions.

In your graph coloring example they would all 3 have the same color meaning they can be merged.

The current algorithm looks at how individual live ranges compare with each other.

Let's say the order in your source code is:
a
b
c

And the pattern rewriter has reached "a".

It will now go through all the candidates i.e. b and c.
When b is analyzed the live range of b (LR(b)) is compared with LR(a). if they are disjoint (assuming all other conditions are met) then b is a valid reuser and it is saved as a valid reuser.

When we then analyze candidate "c" then LR(c) is compared against the LR(a) if it is disjoint it is then compared against the LR(b) (because b is in the list of valid reusers) and if it is disjoint with LR(b) also then and only then is it added as a valid reuser.

In this way it doesn't matter if the order in the code is:

a
b
c

or 

a
c
b

"c" will still be identified as a valid reuser of the same slot as a & b.

Once all candidates have been analyzed and hence all valid reusers collected, the actual replacements are done. At that point the code becomes:
a
a
a

It's only at this point that LR(a) increases.